[Docs] Add docs for custom entity with access per admin channel (#12619)
Sylius/Sylius: docs/cookbook/entities/custom-model-accessible-for-channel-admin.rst

.. rst-class:: plus-doc

How to add a custom model accessible for respective channel administrators?
===========================================================================

Given that you are using Sylius Plus, the licensed edition of Sylius, you may have
the Administrators per Channel defined in your application. Thus when you add a new,
channel-based entity to it, you will need to enable this entity to be viewed only by the relevant channel admins.

1. Define your custom model, our example will be the **Supplier entity**
------------------------------------------------------------------------

In order to prepare a simple Entity follow :doc:`this guide </cookbook/entities/custom-model>`.

> 1. Add your entity to admin menu:
> ----------------------------------
> 
> Adding a new entity to the admin menu is described in the section ``How to customize Admin Menu`` of :doc:`this guide </customization/menu>`.

Reviewer: ```suggestion
Remember to then add your entity to the admin menu. Adding a new entity to the admin menu 
is described in the section ``How to customize Admin Menu`` of :doc:`this guide </customization/menu>`.
```

        }
    }

.. code-block:: yaml

    App\Checker\ResourceChannelEnabilibityChecker:
        decorates: Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelEnabilibityCheckerInterface
        arguments: ['@.inner']

* Create ``App\Checker\ResourceChannelChecker`` and decorate ``Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelCheckerInterface`` next add condition for checking ``Supplier``.

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Checker;

    use Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelCheckerInterface;
    use Sylius\Plus\Entity\ChannelInterface;

    final class ResourceChannelChecker implements ResourceChannelCheckerInterface
    {
        /** @var ResourceChannelCheckerInterface */
        private $decoratedResourceChannelChecker;

        public function __construct(ResourceChannelCheckerInterface $decoratedResourceChannelChecker)
        {
            $this->decoratedResourceChannelChecker = $decoratedResourceChannelChecker;
        }

        public function isFromChannel(object $resource, ChannelInterface $channel): bool
        {
            if (

>                 $resource instanceof Supplier && ($resource->getChannel() === $channel || $resource->getChannel() === null)

Reviewer 2: ```suggestion
                $resource instanceof Supplier && in_array($resource->getChannel(), [$channel, null], true)
```

------------------------------------------------------------------------

In order to prepare a simple Entity follow :doc:`this guide </cookbook/entities/custom-model>`.

* Having your Supplier entity created, add a channel field with relation to the ``Channel`` entity:

.. code-block:: php

    /**
      * @var ChannelInterface
      * @ORM\ManyToOne(targetEntity="Sylius\Plus\Entity\ChannelInterface")
      * @ORM\JoinColumn(name="channel_id", referencedColumnName="id", nullable=true)
      */
     protected $channel;

* Create a proper migration for these changes.

* Next, create a form type for your entity:

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Form;

    use Sylius\Bundle\ResourceBundle\Form\Type\AbstractResourceType;
    use Symfony\Component\Form\Extension\Core\Type\TextType;
    use Symfony\Component\Form\FormBuilderInterface;

    final class SupplierType extends AbstractResourceType
    {
        public function buildForm(FormBuilderInterface $builder, array $options)
        {
            /** @var AdminChannelProviderInterface */
            private $adminChannelProvider;

>             public function __construct(AdminChannelProviderInterface $adminChannelProvider)

Reviewer 3: As this service is an `AbstractResourceType`, shouldn't this constructor look like this:
```suggestion
            public function __construct(string $dataClass, array $validationGroups = [], AdminChannelProviderInterface $adminChannelProvider)
```
? Did you test the cookbook with this implementation?

            {
                $this->adminChannelProvider = $adminChannelProvider;
            }

            $builder
                ->add('name', TextType::class, [
                    'label' => 'Name'
                ])
                ->add('channels', ChannelChoiceType::class, [
                    'choices' => $this->availableChannelsForAdminProvider->getChannels(),
                    'label' => 'sylius.ui.channel',
                ])
            ;
        }

        public function getBlockPrefix(): string
        {
            return 'supplier';
        }
    }

.. code-block:: yaml

>     App\Form\ChannelPerAdminChannelType:

Reviewer 3: ```suggestion
    App\Form\SupplierType:
```

>         arguments: ['@sylius.repository.channel', '@Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface']

Reviewer: ```suggestion
        arguments: [ '@Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface']
```

looking at the constructor it's not needed, did you test it? 🤔

Author: Not needed after fixed this comment: https://github.com/Sylius/Sylius/pull/12619#discussion_r631981376

Reviewer: https://github.com/Sylius/Sylius/pull/12619/files#diff-6f8832a04606b45cd3b3ee77105fdc313fe1e6155bedf622af0c3e17f62a7729R49-R52

Reviewer 3: There is no need for `'@sylius.repository.channel'`, but you should add here `dataClass` and `validationGroups`

>         tags: ['form.type']

Reviewer 3: I guess it should be removed after change to use `ChannelChoiceType`

The ``Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface`` service allows getting a list of proper channels for the currently logged-in admin.

> Remember to register ``App\Form\ChannelPerAdminChannelType`` for resource:

Reviewer 3: ```suggestion
Remember to register ``App\Form\SupplierType`` for resource:
```
?

* Then add ``Supplier`` as checking resource:

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Checker;

    use Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelEnabilibityCheckerInterface;
    use Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelEnabilibityChecker as DecoratedResourceChannelEnabilibityChecker;

    final class ResourceChannelEnabilibityChecker implements ResourceChannelEnabilibityCheckerInterface
    {
        /** @var ResourceChannelEnabilibityCheckerInterface */
        private $decoratedResourceChannelEnabilibityChecker;

        public function __construct(ResourceChannelEnabilibityCheckerInterface $decoratedResourceChannelEnabilibityChecker)
        {
            $this->decoratedResourceChannelEnabilibityChecker = $decoratedResourceChannelEnabilibityChecker;
        }

        public function forResourceName(string $resourceName): bool
        {
            if ($this->decoratedResourceChannelEnabilibityChecker->forResourceName($resourceName)) {
                return true;
            }

>             return in_array($resourceName, ['supplier']);

Reviewer 3: Could be even simpler than my last suggestion:
```suggestion
            return $resourceName === 'supplier';
```

        arguments: ['@.inner']

* Create ``App\Checker\ResourceChannelChecker`` and decorate ``Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelCheckerInterface`` next add condition for checking ``Supplier``.

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Checker;

    use Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelCheckerInterface;
    use Sylius\Plus\Entity\ChannelInterface;

    final class ResourceChannelChecker implements ResourceChannelCheckerInterface
    {
        /** @var ResourceChannelCheckerInterface */
        private $decoratedResourceChannelChecker;

        public function __construct(ResourceChannelCheckerInterface $decoratedResourceChannelChecker)
        {
            $this->decoratedResourceChannelChecker = $decoratedResourceChannelChecker;
        }

        public function isFromChannel(object $resource, ChannelInterface $channel): bool
        {
            // this condition is needed because while creating a Symfony form (our edit/create section), the given object has all fields as null. After creating an empty form this iteration is repeated and properly checked.

>             if ($resource instanceof Supplier && $resource->getChannel() === null) {

Reviewer 3: ```suggestion
            if ($resource instanceof Supplier && $resource->getChannel() === $channel) {
```

    final class SupplierType extends AbstractResourceType
    {
        /** @var AvailableChannelsForAdminProviderInterface */
        private $availableChannelsForAdminProvider;

        public function __construct(
            AvailableChannelsForAdminProviderInterface $availableChannelsForAdminProvider,
            string $dataClass,
            array $validationGroups = []
        ) {
            parent::__construct($dataClass, $validationGroups);

            $this->availableChannelsForAdminProvider = $availableChannelsForAdminProvider;
        }

        public function buildForm(FormBuilderInterface $builder, array $options)
        {
            $builder
                ->add('name', TextType::class, [
                    'label' => 'Name'
                ])
                ->add('channel', ChannelChoiceType::class, [
                    'choices' => $this->availableChannelsForAdminProvider->getChannels(),
                    'label' => 'sylius.ui.channel',
                ])
            ;
        }

        public function getBlockPrefix(): string
        {
            return 'supplier';
        }
    }

.. code-block:: yaml

    App\Form\SupplierType:

>         arguments: ['App\Entity\Supplier', ['sylius'], '@Sylius\Plus\ChannelAdmin\Application\Provider\AvailableChannelsForAdminProviderInterface']

Reviewer: ```suggestion
        arguments: ['@Sylius\Plus\ChannelAdmin\Application\Provider\AvailableChannelsForAdminProviderInterface', 'App\Entity\Supplier', ['sylius']]
```
Shouldn't the arguments here be in the same order as in the class's constructor?

Reviewer 3: IMO, the arguments in constructor should be in different order

        tags: ['form.type']

The ``Sylius\Plus\ChannelAdmin\Application\Provider\AvailableChannelsForAdminProviderInterface`` service allows getting a list of proper channels for the currently logged-in admin.

Remember to register ``App\Form\SupplierType`` for resource:

.. code-block:: yaml

    sylius_resource:
        resources:
            app.supplier:
                driver: doctrine/orm
                classes:
                    model: App\Entity\Supplier
       +            form: App\Form\SupplierType

1. Add your entity to admin menu:
----------------------------------

Adding a new entity to the admin menu is described in the section ``How to customize Admin Menu`` of :doc:`this guide </customization/menu>`.

> 

Reviewer 4: it is part of the preparation for this cookbook so IMHO it should be after the first point (line 13)

> 

Reviewer: Like @AdamKasp says, this whole part should not be a separate point, but should be included into the first point of this cookbook :)

        private $supplierRepository;

        public function __construct(
            AdminChannelProviderInterface $adminChannelProvider,
            EntityRepository $supplierRepository
        ) {
            $this->adminChannelProvider = $adminChannelProvider;
            $this->supplierRepository = $supplierRepository;
        }

        public function create(): QueryBuilder
        {
            $listQueryBuilder = $this->supplierRepository->createQueryBuilder('o');

            /** @var ChannelInterface|null $channel */
            $channel = $this->adminChannelProvider->getChannel();
            if ($channel === null) {
                return $listQueryBuilder;
            }

            return $listQueryBuilder
                ->andWhere('o.channel = :channel')
                ->setParameter('channel', $channel)
            ;
        }
    }

.. code-block:: yaml

    App\Doctrine\ORM\RestrictingSupplierListQueryBuilder:
        public: true
        class: App\Doctrine\ORM\RestrictingSupplierListQueryBuilder
        arguments: ['@Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface', '@app.repository.supplier']

> * Add method to Suppliers grid:

Reviewer: ```suggestion
* Add method to the Suppliers grid:
```

channel-based entity to it, you will need to enable this entity to be viewed only by the relevant channel admins.

1. Define your custom model, our example will be the **Supplier entity**
------------------------------------------------------------------------

In order to prepare a simple Entity follow :doc:`this guide </cookbook/entities/custom-model>`.

* Having your Supplier entity created, add a channel field with relation to the ``Channel`` entity:

.. code-block:: php

    /**
      * @var ChannelInterface
      * @ORM\ManyToOne(targetEntity="Sylius\Plus\Entity\ChannelInterface")
      * @ORM\JoinColumn(name="channel_id", referencedColumnName="id", nullable=true)
      */
     protected $channel;

* Create a proper migration for these changes.

* Next, create a form type for your entity:

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Form;

    use Sylius\Bundle\ResourceBundle\Form\Type\AbstractResourceType;
    use Symfony\Component\Form\Extension\Core\Type\TextType;
    use Symfony\Component\Form\FormBuilderInterface;

    final class SupplierType extends AbstractResourceType
    {

>         /** @var AdminChannelProviderInterface */
>         private $availableChannelsForAdminProvider;
> 
>         /** @var RepositoryInterface */
>         private $channelRepository;
> 
>         public function __construct(
>             string $dataClass,
>             array $validationGroups = [],
>             AdminChannelProviderInterface $availableChannelsForAdminProvider,
>             RepositoryInterface $channelRepository
>         ) {
>             parent::__construct($dataClass, $validationGroups);
> 
>             $this->availableChannelsForAdminProvider = $availableChannelsForAdminProvider;
>             $this->channelRepository = $channelRepository;
>         }
> 
>         public function buildForm(FormBuilderInterface $builder, array $options)
>         {
>             /** @var ChannelInterface $adminChannel */
>             $adminChannel = $this->availableChannelsForAdminProvider->getChannel();
> 
>             $builder
>                 ->add('name', TextType::class, [
>                     'label' => 'Name'
>                 ])
>                 ->add('channel', ChannelChoiceType::class, [
>                     'choices' => $adminChannel === null ? $this->channelRepository->findAll() : [$adminChannel],
>                     'label' => 'sylius.ui.channel',
>                 ])
>             ;
>         }

Reviewer 3: ```suggestion
        /** @var AvailableChannelsForAdminProviderInterface */
        private $availableChannelsForAdminProvider;

        public function __construct(
            string $dataClass,
            array $validationGroups = [],
            AdminChannelProviderInterface $availableChannelsForAdminProvider
        ) {
            parent::__construct($dataClass, $validationGroups);

            $this->availableChannelsForAdminProvider = $availableChannelsForAdminProvider;
        }

        public function buildForm(FormBuilderInterface $builder, array $options)
        {
            $builder
                ->add('name', TextType::class, [
                    'label' => 'Name'
                ])
                ->add('channel', ChannelChoiceType::class, [
                    'choices' => $this->availableChannelsForAdminProvider->getChannels(),
                    'label' => 'sylius.ui.channel',
                ])
            ;
        }
```

        public function getBlockPrefix(): string
        {
            return 'supplier';
        }
    }

.. code-block:: yaml

    App\Form\SupplierType:

>         arguments: ['@Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface']

Reviewer 3: ```suggestion
        arguments: ['@Sylius\Plus\ChannelAdmin\Application\Provider\AvailableChannelsForAdminProviderInterface']
```

        tags: ['form.type']

> The ``Sylius\Plus\ChannelAdmin\Application\Provider\AdminChannelProviderInterface`` service allows getting a list of proper channels for the currently logged-in admin.

Reviewer 3: ```suggestion
The ``Sylius\Plus\ChannelAdmin\Application\Provider\AvailableChannelsForAdminProviderInterface`` service allows getting a list of proper channels for the currently logged-in admin.
```

Adding a new entity to the admin menu is described in the section ``How to customize Admin Menu`` of :doc:`this guide </customization/menu>`.

1. Restrict access to the entity for the respective channel administrator roles (using ACL/RBAC):
-------------------------------------------------------------------------------------------------

* Add access to all resource sections: ``index, create, update, delete, show, bulk_delete``

.. code-block:: yaml

    sylius_plus:
        permissions:
            app_admin_supplier_index:
                parent: suppliers
                label: action.index
            app_admin_supplier_create:
                parent: suppliers
                label: action.create
            app_admin_supplier_update:
                parent: suppliers
                label: action.update
            app_admin_supplier_delete:
                parent: suppliers
                label: action.delete
            app_admin_supplier_show:
                parent: suppliers
                label: action.show
            app_admin_supplier_bulk_delete:
                parent: suppliers
                label: action.bulk_delete

.. note::

>     More information about using administrator roles (ACL/RBAC) can be found :doc:`here </book/customers/admin_user>`.

Reviewer 3: What will happen if we don't do that?

Author: It add actions to permissions tree

Reviewer 4: so it is an RBAC feature. :) is it works without this config? I still think that it is a different feature, but if you really want to mention here about it, I Suggest moving it at the end with a note that it is additional stuff.

        }
    }

.. code-block:: yaml

    App\Checker\ResourceChannelEnabilibityChecker:
        decorates: Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelEnabilibityCheckerInterface
        arguments: ['@.inner']

* Create ``App\Checker\ResourceChannelChecker`` and decorate ``Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelCheckerInterface`` next add condition for checking ``Supplier``.

.. code-block:: php

    <?php

    declare(strict_types=1);

    namespace App\Checker;

    use Sylius\Plus\ChannelAdmin\Application\Checker\ResourceChannelCheckerInterface;
    use Sylius\Plus\Entity\ChannelInterface;

    final class ResourceChannelChecker implements ResourceChannelCheckerInterface
    {
        /** @var ResourceChannelCheckerInterface */
        private $decoratedResourceChannelChecker;

        public function __construct(ResourceChannelCheckerInterface $decoratedResourceChannelChecker)
        {
            $this->decoratedResourceChannelChecker = $decoratedResourceChannelChecker;
        }

        public function isFromChannel(object $resource, ChannelInterface $channel): bool
        {
            if (

>                 ($resource instanceof Supplier && $resource->getChannel() === $channel) ||
>                 $resource instanceof Supplier && $resource->getChannel() === null

Reviewer 3: ```suggestion
                $resource instanceof Supplier && ($resource->getChannel() === $channel) || $resource->getChannel() === null)
```
